packages/core/src/2d/sprite/SpriteRenderer.ts

```diff
@@ -334,16 +350,38 @@ export class SpriteRenderer extends Renderer {
       this._dirtyUpdateFlag &= ~SpriteRendererUpdateFlags.UV;
     }
 
+    // Update color
+    if (this._dirtyUpdateFlag & SpriteRendererUpdateFlags.Color) {
+      this._assembler.updateColor(this);
+      this._dirtyUpdateFlag &= ~SpriteRendererUpdateFlags.Color;
+    }
+
     // Push primitive
-    const texture = this.sprite.texture;
-    const renderData = this._engine._spriteRenderDataPool.getFromPool();
-    renderData.set(this, material, this._verticesData, texture);
-    context.camera._renderPipeline.pushRenderData(context, renderData);
+    const camera = context.camera;
+    const engine = camera.engine;
+    const spriteMaskManager = engine._spriteMaskManager;
+    const renderData = engine._renderData2DPool.get();
+    const chunk = this._chunk;
+    renderData.set(this, material, chunk.data.primitive, chunk.subMesh, this.sprite.texture, chunk);
+    renderData.usage = RenderDataUsage.Sprite;
+    renderData.uploadFlag = ForceUploadShaderDataFlag.None;
+    renderData.preRender = () => {
+      spriteMaskManager.preRender(camera, this);
```

Comment:
- `SpriteMaskManager` needs to be downgraded to a generic design

- `SpriteMaskBatcher` shouldn't exist